### ISSUE TITLE
[EKS/Fargate] [request]: Schedule evenly pod replicas across AZs

### ISSUE THREAD

**Issue by lorenzomicheli** [aws/containers-roadmap#824]
**Tell us about your request**
At the time being, we have observed that there is no guarantee that EKS cluster on Fargate schedule pod replicas evenly across multiple Availability Zones.

**Which service(s) is this request for?**
Fargate, EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
We have a fargate profile with 1 namespace and the 2 private subnets. We have a deployment with 2 replicas and we would like that each replica is deployed on a different AZ. 
At the time being this behaviour is not guaranteed. 

We identified two test cases:
1) Scaling a deployment from 1 replica to 2 replicas: we observed that the 5 times out of 10, the replicas were deployed in the same AZ.

2) Starting from a new deployment configured with 2 replicas: we observed that 3 times out of 10, the replicas were deployed in the same AZ.

**Are you currently working around this issue?**
Currently we are not aware of any workaround.

**Comment by Gowiem** [comment]
@lorenzomicheli and anyone else following this - Has anyone tried using K8s 1.18's [`topologySpreadConstraints`](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/) to deal with this problem? Or will that not work as the Fargate scheduler actually triggers the "node" provision? I'm assuming it does not, but wanted to check before I tried myself 😄

**Comment by lennartt** [comment]
I have the same question as @Gowiem. Unless I'm doing something wrong, the topologySpreadContraints does not seem to be a solution here. Has anyone tried this with success?

**Comment by clemensg** [comment]
Hi! We have a hard requirement of having at least one pod scheduled to each of 3 AZs in a specific region.
Am I right to assume that we have to use managed node groups because topologySpreadConstraints are not (yet?) supported with EKS+Fargate?

**Comment by r0mdau** [comment]
Scheduling evenly is not always pure, but I can recommend using `affinity.podAntiAffinity` as described here https://aws.github.io/aws-eks-best-practices/reliability/docs/application/#schedule-replicas-across-nodes

**Comment by wmosher** [comment]
> Scheduling evenly is not always pure, but I can recommend using `affinity.podAntiAffinity` as described here https://aws.github.io/aws-eks-best-practices/reliability/docs/application/#schedule-replicas-across-nodes

When `topologySpreadConstraints` didn't work as advertised I found this issue and tried last comment.  The `affinity.podAntiAffinity` solution also did not do the trick.  For example, out of 3 available AZs, two scheduled in `b` and two scheduled in `c`, and then scaling to a fifth it was scheduled in `b`.

I would love to hear if there is a plan to get one of these, or some other way to achieve the objective of them, working.

**Comment by foxlollato** [comment]
Hello everyone,
I am having the same issue and wondering about having differents fargate profiles (each one with specific AZ defined) should workout.
Does anyone have tried anything related?

**Comment by foxlollato** [comment]
> Hello everyone, I am having the same issue and wondering about having differents fargate profiles (each one with specific AZ defined) should workout, hence should have a namespace for each AZ as well, or better, create differents fargate profiles regarding same namespace but adding a specific label for the desired AZ. This should work, once You must define the subnets for each fargate profile. Does anyone have tried anything related?

```
Fagate-profile-A
    subnet_id: subnet-id-from-az-a
    podSelectors:
        namespace: my-namespace
        app: my-app
        zone: us-east-1a
    
Fagate-profile-A
    subnet_id: subnet-id-from-az-b
    podSelectors:
        namespace: my-namespace
        app: my-app
        zone: us-east-1b
```
        
Something like this.

**Comment by Youssef-Beltagy** [comment]
There are two current approaches. I recommend the second one. For both approaches, you will have to make a deployment for each AZ.

## Approach 1

1. Prepare subnet1 and subnet2 for the AZ1 and AZ2, respectively.
2. Prepare FP1 with subnet1 and an FP1 selector. 
3. Prepare FP2 with subnet2 and an FP2 selector. This selector must be unique and different from FP1's selector. **Don't use the same selector for FP1 and FP2.**
4. For the replica-set you want in FP1, launch it with the labels for FP1. Do the same for FP2.

Don't leave room for ambiguity on which fargate profile a pod should be mapped to.

## Approach 2 (recommended)

Create a fargate profile with subnets for both AZ1 and AZ2. You can specify which AZ a pod can be matched to in the pod spec. 

```
apiVersion: apps/v1
kind: Deployment
  replicas: 1
  template:
    spec:
      nodeSelector:
        topology.kubernetes.io/zone: us-east-1a
    containers:
        - name: fargate
          image: "nginx"
```

**Comment by marknelissen** [comment]
@Youssef-Beltagy although this solves the "physical" location of the pods, how do you do add stuff like routing, service, autoscaling, etc. on top of these actually separated deployments?

**Comment by visit1985** [comment]
Doesn’t #1125 solve this?

**Comment by WDaan** [comment]
I've found that the following config gives me somewhat the desired behavior:

```
      topologySpreadConstraints:
      - maxSkew: 1 # only 1 pod diff per az
        minDomains: 3 # use at least 3 AZ's
        topologyKey: 'topology.kubernetes.io/zone'
        whenUnsatisfiable: DoNotSchedule
        labelSelector:
          matchLabels:
          ...
```

This way there are at least 3 zones being used, everything scaled above that is still somewhat random and not always perfectly balanced. `ScheduleAnyway` or `DoNotSchedule` doesn't really make a difference.

Any input?

**Comment by jennerm** [comment]
This does not seem to be working for us.  We have the following topologySpreadConstraints on our deployment:
```
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app.kubernetes.io/instance: karpenter
            app.kubernetes.io/name: karpenter
        maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: DoNotSchedule
```
But sometimes we see both fargate nodes in the same AZ:
```
name                                                   az           nodepool   instance-type   arch
fargate-ip-100-80-192-164.eu-west-1.compute.internal   eu-west-1a   <none>     <none>          amd64
fargate-ip-100-80-194-106.eu-west-1.compute.internal   eu-west-1a   <none>     <none>          amd64
```
I looked at https://github.com/aws/containers-roadmap/issues/1125 referenced above, but it only seems to talk about using nodeSelector or nodeAffinity rather than than topologySpreadConstraints.

**Comment by mapu77** [comment]
Using common k8s pod topology won't work in EKS Fargate as stated clearly in their docs.

> Note
Fargate does not currently support Kubernetes [topologySpreadConstraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/)
_(from https://docs.aws.amazon.com/en_us/eks/latest/userguide/fargate-profile.html)_

In the same piece of documentation, they suggest a possible solution ([following @Youssef-Beltagy approach 1](https://github.com/aws/containers-roadmap/issues/824#issuecomment-1883718392))

> If you must have an even spread, use two Fargate profiles. Even spread is important in scenarios where you want to deploy two replicas and don’t want any downtime. We recommend that each profile has only one subnet.
_(from https://docs.aws.amazon.com/en_us/eks/latest/userguide/fargate-profile.html)_

However, I don't think this solution scales up; this must be included as part of the Fargate feature offering

**Comment by Youssef-Beltagy** [comment]
I'm not a member of EKS team at this moment, so I'm not speaking on behalf of AWS/EKS.

TLDR: use **[EKS Auto](https://docs.aws.amazon.com/eks/latest/userguide/auto-migrate-fargate.html)**

---

@visit1985, @WDaan, @jennerm
EKS Fargate doesn't support pod topology constraints (as of right now). As @mapu77 said, this is written in the public docs. The way this works is that the deployment request will not fail, but the topology spread constraints will not be honored: the topology spread constraints will be ignored.

As for what @WDaan saw:
1. Maybe @WDaan didn't confirm the pods were spread across multiple AZs.
2. The Fargate scheduler puts pods in the subnets with most available IPs (IIRC). If you have symmetric subnets, this could appear as topology spread. Please don't intentionally use this approach for multi-AZ deployment: it can easily break.
3. Maybe it was just that the stars aligned right for @WDaan

As far as I'm aware, the [two approaches I mentioned](https://github.com/aws/containers-roadmap/issues/824#issuecomment-1883718392) are the recommendations from AWS to use multi-AZ deployment with EKS Fargate. @mapu77 and @marknelissen are right in that both approaches are problematic as you have to scale multiple deployments (though I don't believe it is problematic for selecting pods for services). But I think it is unlikely for EKS Fargate to take up this feature as they are driving EKS Auto as the future of EKS managed compute.

**[The recommendation from EKS is to migrate to EKS Auto](https://docs.aws.amazon.com/eks/latest/userguide/auto-migrate-fargate.html)**.

**Comment by mapu77** [comment]
I don't understand why is unlikely that EKS Fargate offers this feature, having [ECS supporting even AZ distribution](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-rebalancing.html) @Youssef-Beltagy?

To me every solution provided by AWS gives a different degree of control over the cluster. IMO EKS Fargate represents the ultimate low-configuration environment within AWS EKS offering (one just needs to describe profile rules - i.e. namespace - and indicate application resources to have something running).

With EKS Auto, IMO a higher degree of configuration is required when we need to configure nodepools resources (inherited from karpenter configuration).

**Comment by visit1985** [comment]
In the last years Fargate didn't receive much love by AWS - thinking about this issue or #2274.
Using two Fargate profiles was never an option for a Fargate-only cluster with one wildcard Fargate profile.
Finally AWS forced me to move away from it, by deprecating AWS App Mesh and not providing any alternative (#682, #1027).

**Comment by alegriaw** [comment]
Hi, do we have any new progress or feature update for this issue on Fargate?
